pkg/network/ebpf/c/protocols/read_into_buffer.h

```diff
@@ -0,0 +1,51 @@
+#define READ_INTO_BUFFER(name, total_size, blk_size)                                                                \
+    static __always_inline void read_into_buffer_##name(char *buffer, struct __sk_buff *skb, skb_info_t *info) {    \
+        u64 offset = (u64)info->data_off;                                                                           \
+        const u32 len = (total_size) < (skb->len - info->data_off) ? info->data_off + (total_size) : skb->len;      \
+        unsigned i = 0;                                                                                             \
+                                                                                                                    \
+    _Pragma( STRINGIFY(unroll(total_size/blk_size)) )                                                               \
+        for (; i < ((total_size) / (blk_size)); i++) {                                                              \
+            if (offset + (blk_size) - 1 >= len) { break; }                                                          \
+                                                                                                                    \
+            bpf_skb_load_bytes_with_telemetry(skb, offset, &buffer[i * (blk_size)], (blk_size));                    \
+            offset += (blk_size);                                                                                   \
+        }                                                                                                           \
+                                                                                                                    \
+        const s64 left_payload = (s64)len - (s64)offset;                                                            \
+        if (left_payload < 1) {                                                                                     \
+            return;                                                                                                 \
+        }                                                                                                           \
+                                                                                                                    \
+        s64 read_size = (blk_size) - 1;                                                                             \
+        if (left_payload < read_size) {                                                                             \
+            read_size = left_payload;                                                                               \
+        }                                                                                                           \
+                                                                                                                    \
+        const s64 left_buffer = (s64)CLASSIFICATION_MAX_BUFFER - (s64)(i*(blk_size));                               \
```

Comment:
I'm a bit confused about all of these conditionals :thinking: 

First, why is  `CLASSIFICATION_MAX_BUFFER` being referenced here and how does that relate to `total_size`?

My understanding is:
1. Your code will read _up to_ `total_size` bytes in chunks of `blk_size`;
2. The remaining read should be `(len-offset)%blk_size` so I don't get why checking for `left_payload < read_size` is necessary
3. Finally, how would `read_size` be greater than `left_buffer`? I think this check is ensuring you wouldn't read more than your buffer capacity, but I don't understand how this would be possible.

If these checks are only here for the purposes of appeasing the verifier somehow could you please make that clear by adding comments?

Reply:
> First, why is CLASSIFICATION_MAX_BUFFER being referenced here and how does that relate to total_size?

good catch, indeed that's a bug and should be CLASSIFICATION_MAX_BUFFER

regarding the other 2 bullets, indeed you are getting it right, but the verifier does not approve
```
        const s64 left_payload = (s64)len - (s64)offset;                                                            \
        if (left_payload < 1) {                                                                                     \
            return;                                                                                                 \
        }                                                                                                           \
                                                                                                                    \
        const s64 reminder = left_payload % (blk_size);                                                             \
        if (0 < reminder && reminder < blk_size) {                                                                  \
            bpf_skb_load_bytes_with_telemetry(skb, offset, &buffer[i * (blk_size)], reminder);                      \
        }                                                                                                           \
        return;                                                                                                     \
```

(or any other equivalent)
as it complains we have unbounded memory access

I'll add comments clarifying the code